Source/CarthageKit/Algorithms.swift

```diff
@@ -65,3 +65,38 @@ public func topologicalSort<Node: Comparable>(graph: Dictionary<Node, Set<Node>>
 
 	return workingGraph.isEmpty ? sorted : nil
 }
+
+/// Performs a topological sort on the provided graph with its output sorted to
+/// include only the provided set of nodes and their transitively incoming 
+/// nodes (dependencies).
+///
+/// Returns nil if any of the provided nodes are not contained within the
+/// provided graph or if the provided graph has a cycle.
+public func topologicalSort<Node: Comparable>(graph: Dictionary<Node, Set<Node>>, nodes: Set<Node>) -> [Node]? {
+	guard !nodes.isEmpty else { return topologicalSort(graph) }
+
+	guard nodes.isSubsetOf(Set(graph.keys)) else { return nil }
```

Comment:
It seems like this should be a `precondition` instead of a `guard`. It should indicate an error in Carthage's logic, rather than an error in `Cartfile`s, I think.